### PR TITLE
[Form] deprecate not using a rounding mode

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | yes
| Tickets       | 
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by xabbuh** [comment]
I didn't manage to look at #35729 before and did it now. But I think it contains some BC breaks that we need to cover with proper deprecations first which I did here.

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/PercentToLocalizedStringTransformer.php [hunk 1]
```diff
@@ -103,8 +102,8 @@ public function __construct(int $scale = null, string $type = null, ?int $roundi
             $type = self::FRACTIONAL;
         }
 
-        if (null === $roundingMode) {
-            $roundingMode = self::ROUND_HALF_UP;
+        if (null === $roundingMode && (\func_num_args() < 4 || func_get_arg(3))) {
+            trigger_deprecation('symfony/form', '5.1', sprintf('Not passing a rounding mode to %s() is deprecated. Starting with Symfony 6.0 it will default to %s::ROUND_HALF_UP.', __METHOD__, __CLASS__));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/PercentToLocalizedStringTransformer.php:106]: missing double quotes in the message

File: src/Symfony/Component/Form/Extension/Core/Type/PercentType.php [hunk 2]
```diff
@@ -48,7 +49,11 @@ public function configureOptions(OptionsResolver $resolver)
     {
         $resolver->setDefaults([
             'scale' => 0,
-            'rounding_mode' => NumberToLocalizedStringTransformer::ROUND_HALF_UP,
+            'rounding_mode' => function (Options $options) {
+                trigger_deprecation('symfony/form', '5.1', sprintf('Not configuring the "rounding_mode" option is deprecated. It will default to %s::ROUND_HALF_UP in Symfony 6.0.', PercentToLocalizedStringTransformer::class));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Core/Type/PercentType.php:53]: double quotes also?

File: src/Symfony/Component/Form/Extension/Core/Type/PercentType.php [hunk 3]
```diff
@@ -59,16 +64,24 @@ public function configureOptions(OptionsResolver $resolver)
...
+            PercentToLocalizedStringTransformer::ROUND_HALF_EVEN,
+            PercentToLocalizedStringTransformer::ROUND_HALF_UP,
+            PercentToLocalizedStringTransformer::ROUND_UP,
+            PercentToLocalizedStringTransformer::ROUND_CEILING,
         ]);
         $resolver->setAllowedTypes('scale', 'int');
         $resolver->setAllowedTypes('symbol', ['bool', 'string']);
+        $resolver->setDeprecated('rounding_mode', 'symfony/form', '5.1', function (Options $options, $roundingMode) {
+            if (null === $roundingMode) {
+                return sprintf('Not configuring the "rounding_mode" option is deprecated. It will default to %s::ROUND_HALF_UP in Symfony 6.0.', PercentToLocalizedStringTransformer::class);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Core/Type/PercentType.php:80]: double quotes

**Review by nicolas-grekas — COMMENTED** [review]
What's the current default (before #35729)? Why change it ?

**Comment by xabbuh** [comment]
In 4.4 the transformer will simply not round and act like in this test: https://github.com/symfony/symfony/pull/36578/files#diff-3d096748c112b424e758960eb06177caR40

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Pull request merged**